athena-storage-api/src/main/java/com/amazonaws/athena/storage/StorageUtil.java

```diff
@@ -0,0 +1,101 @@
+/*-
+ * #%L
+ * Amazon Athena Storage API
+ * %%
+ * Copyright (C) 2019 - 2022 Amazon Web Services
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package com.amazonaws.athena.storage;
+
+import com.google.common.base.CharMatcher;
+import com.univocity.parsers.common.routine.InputDimension;
+import com.univocity.parsers.csv.CsvRoutines;
+
+import java.io.File;
+import java.util.Locale;
+
+public class StorageUtil
+{
+    /**
+     * Regular expression to remove invalid characters from entity name as per ANSI/ISO Entry SQL-92 standard, except
+     * entity names in GCS connector are all lower case
+     * Also replace invalid characters for file naming (for example, '/' characters is not valid for file naming)
+     */
+    private static final String INVALID_CHARS_REPLACE_REGEX = "[^A-Za-z0-9]+|[\\\\/:*?\"<>|]+";
+
+    /**
+     * Regular expression to see whether a string starts with number(s)
+     */
+    private static final String STARTS_DIGIT_REGEX = "^\\d+";
+
+    /**
+     * Character matchers to retain only alphanumeric characters and an underscore
+     * It helps to replace any special characters from the string
+     */
+    private static final CharMatcher ALPHA_NUM_MATCHER =
+            CharMatcher.inRange('a', 'z')
+                    .or(CharMatcher.inRange('A', 'Z'))
+                    .or(CharMatcher.inRange('0', '9')).precomputed()
+                    .or(CharMatcher.is('_'));
+
+    private StorageUtil()
+    {
+    }
+
+    /**
+     * Counts the CSV records with the help of {@link CsvRoutines#getInputDimension(File)}
+     *
+     * @return Record count in the CSV file
+     */
+    public static synchronized long getCsvRecordCount(File cachedFile)
+    {
+        InputDimension dimension = new CsvRoutines().getInputDimension(cachedFile);
+        // excluding the header row
+        return dimension.rowCount() - 1;
+    }
```

Comment:
Yes. This is because the CSV parser reads data byte by byte. Reading each byte from the storage stream dramatically slows down the record fetching. Which is why we cache the entire bytes in the ephemeral storage. As @kparwal suggested, as there is a constraint of 500 MB size on CSV, the data source should be able read multiple files, much like seen in hive meta store format.

We'll refactor it during meta-store-like partitioning implementation

Reply:
Now part of partitionig